### DISCUSSION TITLE
Is there a way to get caret position?

### DISCUSSION THREAD

**Discussion by maraf** [microsoft/terminal#14664] [Q&A] [answered]
I'm building an application that opens a small window next to caret, and so I need to find a position of the caret.

In some applications, using `IAccessible.accLocation` provides good results. 
In other ones, using `IUIAutomationTextPattern` or `IUIAutomationTextPattern2` provides good results.
Sometimes Win32 `GetCaretPos` works great.

None of these seem to be working for Terminal. 
Is there a way to get caret position?

**Comment by maraf** [comment]
I have a progress. I'm able to get a `IUIAutomationTextPattern`, but still can't get the caret position.

(the code assumes that terminal has focus)

```c#
var automation = new CUIAutomation();
var element = automation.GetFocusedElement();
var pattern = (IUIAutomationTextPattern)element.GetCurrentPattern(10014);
if (pattern != null)
{
    var selectons = pattern.GetSelection();
    if (selectons.Length > 0)
    {
        var selection = selectons.GetElement(0);
        var bounds = selection.GetBoundingRectangles();
        // bounds is an empty array 👈
    }
}
```

Is this a good path? Should it work? Or am I missing something?

**Reply by j4james** [reply]
I haven't played with the `UIAutomation` functionality myself, but just based on what I've seen of the implementation in the Windows Terminal source, I think you're on the right track. As far as I can see, `IUIAutomationTextPattern2` is not supported, so `GetCaretRange` isn't an option, but `GetSelection` should return the cursor position when nothing else is selected. See here:

https://github.com/microsoft/terminal/blob/4d9a266c12d6bdafa88476279e5761f5eb5ff4fc/src/types/ScreenInfoUiaProviderBase.cpp#L237-L243

I think your problem is that the returned selection is then going to be zero width, so that would explain why there are no bounding rectangles. And this is a wild guess, but you could try using `ExpandToEnclosingUnit` to normalize the range to one character unit, and then see what `GetBoundingRectangles` returns.

Otherwise, I've also seen some people saying you can use `CompareEndpoints` to compare the range against something like the document range, and that'll give you an offset into the document. However, I don't think that's guaranteed to work, and you've still then got to figure out the 2D coordinates from that somehow.

**Reply by maraf** [reply]
👏 Thank you! The `ExpandToEnclosingUnit` made the trick. Now bounds contain cursor position.
If post it as a separate answer, I can mark as THE answer to get you kudos 🙂

**Comment by j4james** [comment, marked as answer]
To summarise what we've established in the other threads:

1. Windows Terminal doesn't support `IUIAutomationTextPattern2`, but the `GetSelection` method from `IUIAutomationTextPattern` should return the caret position when nothing else is selected.
2. However, that returned selection will have a zero width, so you'll then need to call `ExpandToEnclosingUnit` to normalize the range to one character unit.
3. Once the range has a non-zero width, you can call `GetBoundingRectangles` to obtain the boundaries, and that should give you the position you're looking for.
